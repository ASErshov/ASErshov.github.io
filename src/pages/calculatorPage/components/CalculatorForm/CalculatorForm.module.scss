@import "../../../../mixins/breakpoints.scss";

.mainWrapper {
  display: flex;
  flex-direction: column;
  height: 100%;
  width: 100%;
  padding: 0 10px 0 10px;
  box-sizing: border-box;
}

.input {
  direction: rtl;
  caret-color: transparent;
  border: none;

  &::placeholder {
    direction: rtl;
  }

  &::-webkit-outer-spin-button,
  &::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }
  -moz-appearance: textfield;
}

.inputWrapper {
  &:not(:hover) fieldset {
    border-width: 0;
  }
  &:hover input:focus + fieldset {
    border-width: 2px;
  }
}

.label.label {
  &:not(:first-child) {
    margin: 30px 0 6px 0;
  }
  &:first-child {
    margin: 0 0 6px 0;
  }
}

.tipWrapper {
  display: flex;
  flex-wrap: wrap;
  column-gap: 18px;
  row-gap: 16px;
  margin-top: 10px;
}

.tipButton.tipButton {
  width: calc(50% - 9px);
  height: 48px;
  box-sizing: border-box;
}

.peopleLableWrapper {
  display: flex;
  justify-content: space-between;
}

.error fieldset {
  border: 2px solid red !important;
}

@include breakpoint("laptopLandscape") {
  .mainWrapper {
    padding: 0;
    width: 380px;
    margin: 16px;
  }

  .tipWrapper {
    column-gap: 14px;
    row-gap: 16px;
    margin-top: 16px;
  }

  .tipButton.tipButton {
    width: 117px;
    height: 48px;
  }

  .label.label {
    margin: 40px 0 2px 0;
  }
}
