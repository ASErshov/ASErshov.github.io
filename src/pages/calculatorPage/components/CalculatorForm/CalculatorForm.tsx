import React, { useEffect, useRef, useState } from "react";
import { Box, Button, InputAdornment, OutlinedInput, Typography } from "@mui/material";
import { ReactComponent as DollarSvg } from "assets/icon-dollar.svg";
import { ReactComponent as PersonSVG } from "assets/icon-person.svg";
import { CalculatorValues } from "pages/calculatorPage/types";
import styles from "./CalculatorForm.module.scss";

interface Props {
  calculate: (value: CalculatorValues) => void;
  error?: boolean;
  isReset?: boolean;
}

const CalculatorForm: React.FC<Props> = ({ calculate, error, isReset }) => {
  const tips = [5, 10, 15, 25, 50];

  const submitRef = useRef<HTMLInputElement>(null);
  const [tip, setTip] = useState<number>(0);
  const [customTip, setCustomTip] = useState<string | undefined>("");
  const [peopleNumber, setPeopleNumber] = useState<string>("");
  const [bill, setBill] = useState<string>("");

  useEffect(() => {
    if (isReset) {
      setCustomTip("");
      setTip(0);
      setPeopleNumber("");
      setBill("");
    }
  }, [isReset]);

  const handleChange = (): void => {
    submitRef.current?.click();
  };

  const handleTipClick = async (value: number): Promise<void> => {
    await setTip(value);
    setCustomTip("");
    handleChange();
  };

  const handleChangeCustomTip = async ({ target: { value } }: { target: { value: string } }): Promise<void> => {
    await setTip(Number(value));
    setCustomTip(value);
    handleChange();
  };

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>): void => {
    e.preventDefault();
    const formData = new FormData(e.target as HTMLFormElement);
    const formProps = Object.fromEntries(formData);
    calculate({ ...formProps, tip });
  };

  return (
    <form className={styles.mainWrapper} onSubmit={handleSubmit}>
      <Typography className={styles.label} variant="body2" component="span" sx={{ color: "primary.dark" }}>
        Bill
      </Typography>
      <OutlinedInput
        id="bill"
        name="bill"
        data-testid="bill"
        placeholder="0"
        startAdornment={
          <InputAdornment position="start">
            <DollarSvg />
          </InputAdornment>
        }
        inputProps={{
          className: styles.input,
          sx: {
            color: "secondary.dark",
          },
        }}
        className={styles.inputWrapper}
        size="small"
        type="number"
        sx={{
          backgroundColor: "secondary.light",
        }}
        onChange={({ target: { value } }): void => {
          setBill(value);
          handleChange();
        }}
        value={bill}
      />
      <Typography className={styles.label} variant="body2" component="span" sx={{ color: "primary.dark" }}>
        Select Tip %
      </Typography>
      <Box className={styles.tipWrapper}>
        {tips.map((item) => (
          <Button
            className={styles.tipButton}
            key={item}
            onClick={(): void => {
              handleTipClick(item);
            }}
            sx={{
              color: `${tip === item ? "secondary.dark" : "common.white"}`,
              backgroundColor: `${tip === item ? "primary.main" : "secondary.dark"}`,
              "&:hover": {
                color: "secondary.dark",
                backgroundColor: "primary.main",
              },
            }}
            variant="contained"
          >{`${item}%`}</Button>
        ))}
        <OutlinedInput
          id="customTip"
          name="customTip"
          data-testid="customTip"
          placeholder="Custom"
          className={`${styles.inputWrapper}  ${styles.tipButton}`}
          inputProps={{
            className: styles.input,
            sx: {
              "&::placeholder": {
                color: "secondary.dark",
              },
              color: "secondary.dark",
            },
          }}
          size="small"
          type="number"
          sx={{
            backgroundColor: "secondary.light",
          }}
          onChange={handleChangeCustomTip}
          value={customTip}
        />
      </Box>
      <div className={`${styles.peopleLableWrapper} ${styles.label}`}>
        <Typography variant="body2" component="span" sx={{ color: "primary.dark" }}>
          Number of people
        </Typography>
        {error && (
          <Typography variant="body2" component="span" sx={{ color: "red" }}>
            Can't be zero
          </Typography>
        )}
      </div>
      <OutlinedInput
        id="peopleNumber"
        name="peopleNumber"
        data-testid="peopleNumber"
        placeholder="0"
        startAdornment={
          <InputAdornment position="start">
            <PersonSVG />
          </InputAdornment>
        }
        className={`${styles.inputWrapper} ${error ? styles.error : ""}`}
        inputProps={{
          className: styles.input,
          sx: {
            color: "secondary.dark",
          },
        }}
        size="small"
        type="number"
        sx={{
          backgroundColor: "secondary.light",
        }}
        onChange={({ target: { value } }): void => {
          setPeopleNumber(value);
          handleChange();
        }}
        value={peopleNumber}
      />
      <input style={{ display: "none" }} type="submit" ref={submitRef} />
    </form>
  );
};

export default CalculatorForm;
