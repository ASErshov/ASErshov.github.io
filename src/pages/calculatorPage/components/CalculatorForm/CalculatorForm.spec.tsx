import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import CalculatorForm from "./CalculatorForm";
import renderer from "react-test-renderer";

/* eslint-disable @typescript-eslint/explicit-function-return-type */

it("Matches DOM Snapshot", () => {
  const MOCKED_CALCULATE = jest.fn();
  const domTree = renderer.create(<CalculatorForm calculate={MOCKED_CALCULATE} />).toJSON();
  expect(domTree).toMatchSnapshot();
});

test("renders the CalculatorForm", () => {
  const MOCKED_CALCULATE = jest.fn();
  render(<CalculatorForm calculate={MOCKED_CALCULATE} />);
  expect(screen.getByText("Bill")).toBeInTheDocument();
  expect(screen.getByText("Select Tip %")).toBeInTheDocument();
  expect(screen.getByText("Number of people")).toBeInTheDocument();
});

test("render with isRest", () => {
  const MOCKED_CALCULATE = jest.fn();
  render(<CalculatorForm calculate={MOCKED_CALCULATE} isReset={true} />);
  expect(screen.getByTestId("bill").getElementsByTagName("input").item(0)).toHaveDisplayValue("");
  expect(screen.getByTestId("customTip").getElementsByTagName("input").item(0)).toHaveDisplayValue("");
  expect(screen.getByTestId("peopleNumber").getElementsByTagName("input").item(0)).toHaveDisplayValue("");
});

test("render with error", () => {
  const MOCKED_CALCULATE = jest.fn();
  render(<CalculatorForm calculate={MOCKED_CALCULATE} error={true} />);
  expect(screen.getByText("Can't be zero")).toBeInTheDocument();
});

test("change functions test", async () => {
  const MOCKED_CALCULATE = jest.fn();
  render(<CalculatorForm calculate={MOCKED_CALCULATE} />);

  const billInput = screen.getByTestId("bill").getElementsByTagName("input").item(0);
  billInput && fireEvent.change(billInput, { target: { value: "123" } });
  expect(billInput).toHaveDisplayValue("123");
  expect(MOCKED_CALCULATE.mock.calls.length).toBe(1);

  const tipButton = screen.getByText("10%");
  tipButton && fireEvent.click(tipButton, {});
  await waitFor(() => expect(MOCKED_CALCULATE.mock.calls.length).toBe(2));

  const customTipInput = screen.getByTestId("customTip").getElementsByTagName("input").item(0);
  customTipInput && fireEvent.change(customTipInput, { target: { value: "123" } });
  await waitFor(() => expect(customTipInput).toHaveDisplayValue("123"));

  expect(MOCKED_CALCULATE.mock.calls.length).toBe(3);
  customTipInput && fireEvent.change(customTipInput, { target: { value: "" } });
  await waitFor(() => expect(customTipInput).toHaveDisplayValue(""));
  expect(MOCKED_CALCULATE.mock.calls.length).toBe(4);

  const peopleNumberInput = screen.getByTestId("peopleNumber").getElementsByTagName("input").item(0);
  peopleNumberInput && fireEvent.change(peopleNumberInput, { target: { value: "123" } });
  expect(peopleNumberInput).toHaveDisplayValue("123");
  expect(MOCKED_CALCULATE.mock.calls.length).toBe(5);
});
