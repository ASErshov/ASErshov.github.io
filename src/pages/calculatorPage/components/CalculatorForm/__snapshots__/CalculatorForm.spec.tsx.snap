// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches DOM Snapshot 1`] = `
<form
  className="mainWrapper"
  onSubmit={[Function]}
>
  <span
    className="MuiTypography-root MuiTypography-body2 label css-htvaao-MuiTypography-root"
  >
    Bill
  </span>
  <div
    className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-sizeSmall MuiInputBase-adornedStart inputWrapper css-kgyakk-MuiInputBase-root-MuiOutlinedInput-root"
    data-testid="bill"
    onClick={[Function]}
  >
    <div
      className="MuiInputAdornment-root MuiInputAdornment-positionStart css-ittuaa-MuiInputAdornment-root"
    >
      <span
        className="notranslate"
      >
        ​
      </span>
      <svg>
        icon-dollar.svg
      </svg>
    </div>
    <input
      className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedStart input css-1pe19lk-MuiInputBase-input-MuiOutlinedInput-input"
      id="bill"
      name="bill"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="0"
      type="number"
      value=""
    />
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-ihdtdm"
      >
        <span
          className="notranslate"
        >
          ​
        </span>
      </legend>
    </fieldset>
  </div>
  <span
    className="MuiTypography-root MuiTypography-body2 label css-htvaao-MuiTypography-root"
  >
    Select Tip %
  </span>
  <div
    className="tipWrapper MuiBox-root css-0"
  >
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium tipButton css-1n3tt0b-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      5%
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium tipButton css-1n3tt0b-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      10%
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium tipButton css-1n3tt0b-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      15%
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium tipButton css-1n3tt0b-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      25%
    </button>
    <button
      className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium tipButton css-1n3tt0b-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      50%
    </button>
    <div
      className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-sizeSmall inputWrapper  tipButton css-1whsbrh-MuiInputBase-root-MuiOutlinedInput-root"
      data-testid="customTip"
      onClick={[Function]}
    >
      <input
        className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall input css-19s102u-MuiInputBase-input-MuiOutlinedInput-input"
        id="customTip"
        name="customTip"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Custom"
        type="number"
        value=""
      />
      <fieldset
        aria-hidden={true}
        className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          className="css-ihdtdm"
        >
          <span
            className="notranslate"
          >
            ​
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
  <div
    className="peopleLableWrapper label"
  >
    <span
      className="MuiTypography-root MuiTypography-body2 css-htvaao-MuiTypography-root"
    >
      Number of people
    </span>
  </div>
  <div
    className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-sizeSmall MuiInputBase-adornedStart inputWrapper  css-kgyakk-MuiInputBase-root-MuiOutlinedInput-root"
    data-testid="peopleNumber"
    onClick={[Function]}
  >
    <div
      className="MuiInputAdornment-root MuiInputAdornment-positionStart css-ittuaa-MuiInputAdornment-root"
    >
      <span
        className="notranslate"
      >
        ​
      </span>
      <svg>
        icon-person.svg
      </svg>
    </div>
    <input
      className="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedStart input css-1pe19lk-MuiInputBase-input-MuiOutlinedInput-input"
      id="peopleNumber"
      name="peopleNumber"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder="0"
      type="number"
      value=""
    />
    <fieldset
      aria-hidden={true}
      className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        className="css-ihdtdm"
      >
        <span
          className="notranslate"
        >
          ​
        </span>
      </legend>
    </fieldset>
  </div>
  <input
    style={
      Object {
        "display": "none",
      }
    }
    type="submit"
  />
</form>
`;
